### PR TITLE
Allow json reader/decoder to work with format_strings for each field 

### PR DESCRIPTION
**Pull request by sum12**
# Which issue does this PR close?
add implementation for remaining basic types as needed by the json::decoder

# Rationale for this change
This approach is seem cleaner than the on e initially proposed in #1301 

# What changes are included in this PR?
Non-Formatted parsing for timestamp types is included, but formatted parsing is still missing and can be done as a followup PR.
When `serde` thinks that the `value` is a `string` then these new parsers will kick in and parse the string using provided format  or some predefined set of format strings (for timestamp types)
for time32/time64 based types no format is specified

# Are there any user-facing changes?
A new `format_strings` parameter is added to the reader and decoder

### CODE REVIEW

**Comment by alamb** [comment]
Will try and check this out tomorrow

**Review by alamb — COMMENTED** [review]
Thanks @sum12  -- sorry for the delayed review

File: arrow/src/json/reader.rs [hunk 1]
```diff
@@ -589,11 +592,13 @@ impl Decoder {
         schema: SchemaRef,
         batch_size: usize,
         projection: Option<Vec<String>>,
+        format_strings: Option<HashMap<String, String>>,
```
[the remark this answers is not in the document]
Comment (sum12) [reply arrow/src/json/reader.rs:595]: I often wondered too. Wouldn't it allow the parser to diverge from schema ? We would need a runtime validation of what schema says and formaters actually return.

File: arrow/src/json/reader.rs [hunk 2]
```diff
@@ -38,7 +38,7 @@
 //!
 //! let file = File::open("test/data/basic.json").unwrap();
 //!
-//! let mut json = json::Reader::new(BufReader::new(file), Arc::new(schema), 1024, None);
+//! let mut json = json::Reader::new(BufReader::new(file), Arc::new(schema), 1024, None, None);
```
[the remark this answers is not in the document]
Comment (sum12) [reply arrow/src/json/reader.rs]: added a `DecoderOptions` as decoder is like to get those new options in the future than the reader.

**Comment by sum12** [comment]
@alamb re-review please :-)

File: arrow/src/util/reader_parser.rs [hunk 3]
```diff
@@ -60,27 +60,39 @@ impl Parser for Int8Type {}
...
+}
+
+impl Parser for TimestampSecondType {
+    fn parse(string: &str) -> Option<i64> {
+        let nanos = string_to_timestamp_nanos(string).ok();
+        nanos.map(|x| x / 1_000_000_000)
     }
 }
 
+impl Parser for Time64NanosecondType {}
```
Comment (alamb) [inline arrow/src/util/reader_parser.rs:88]: If we do this, doesn't it mean that integers will be parsed as time64s? Is that what you want to allow? If so can you please explain / write a test that covers the behavior (so we don't accidentally break it in the future?)

File: arrow/src/json/reader.rs [hunk 4]
```diff
@@ -577,30 +579,34 @@ pub struct Decoder {
     /// Explicit schema for the JSON file
     schema: SchemaRef,
     /// Optional projection for which columns to load (case-sensitive names)
```
[the remark this answers is not in the document]
Comment (alamb) [reply arrow/src/json/reader.rs]: I fixed it in 8427ce6e52

**Review by alamb — APPROVED** [review]
I am sorry for the delay in reviewing @sum12 

I think this one is almost ready to go. Thank you for pulling out out the options

I merged to master and pushed a commit to fix the comments

File: arrow/src/util/reader_parser.rs (hunk 3 above)
Comment (sum12) [reply arrow/src/util/reader_parser.rs:88]: thanks for pointing that out. Have added tests to parse string from json files to these types. iiuc this latest commit could be also cherry picked to your branch `alamb:alamb/batch_size_too` ? and this one can be closed.

**Review by alamb — APPROVED** [review]

**Comment by alamb** [comment]
Looks good -- thanks @sum12

**Pull request merged**